framework/src/executioners/Transient.C

```diff
@@ -228,33 +228,36 @@ Transient::postStep()
 void
 Transient::execute()
 {
-
   preExecute();
 
+  if (_app.isRecovering())
+    incrementStepOrReject();
+
   // NOTE: if you remove this line, you will see a subset of tests failing. Those tests might have a wrong answer and might need to be regolded.
   // The reason is that we actually move the solution back in time before we actually start solving (which I think is wrong).  So this call here
   // is to maintain backward compatibility and so that MOOSE is giving the same answer.  However, we might remove this call and regold the test
   // in the future eventually.
-  if (!_app.isRecovering())
+  else
     _problem.advanceState();
 
+
   // Start time loop...
-  while (true)
+  while (keepGoing())
   {
-    if (_first != true)
-      incrementStepOrReject();
-
+#ifdef LIBMESH_ENABLE_AMR
+    if (!_first && _problem.adaptivity().isOn())
+      _problem.adaptMesh();
+#endif
```

Comment:
If somebody needed `time` during adaptivity, like in markers and indicators, they get an incorrect value if we do it this way. Time is already advanced at this point.  Not sure how big issue this is, but just that you are aware of it...

May be the logic should be:
```
solve step
decide if we need to continue
if so, do
  adaptivity (if on)
  advance time step
```